### PR TITLE
Update CollectibleContract data when it changes

### PR DESCRIPTION
**Pull request by adonesky1**
- CHANGED:

  - Modifies the `addCollectibleContract` on the `CollectiblesController` to update `CollectibleContract` data if there is a previous entry in state that matches the address of the contract currently being processed but other fields that don't match.

### CODE REVIEW

File: src/assets/assetsUtil.ts [hunk 1]
```diff
@@ -34,3 +38,40 @@ export function compareCollectiblesMetadata(
...
+    'createdDate',
+    'description',
+    'externalLink',
+    'logo',
+    'name',
+    'schemaName',
+    'symbol',
+    'totalSupply',
+  ];
+  const differentValues = keys.reduce((value, key) => {
```
Comment (wachunei) [inline src/assets/assetsUtil.ts:67]: Maybe it is better we exit as soon as we find a different value using [`keys.some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)

Comment (adonesky1) [reply src/assets/assetsUtil.ts]: done here: https://github.com/MetaMask/controllers/pull/680/commits/1be1ed0c058f09d9483d0c93d233657ed117cffd

File: src/assets/assetsUtil.ts [hunk 2]
```diff
@@ -34,3 +38,34 @@ export function compareCollectiblesMetadata(
...
+/**
+ * Compares CollectibleContract entries to any CollectibleContract entry.
+ * We need this method when comparing a new fetched CollectibleContract, in case a entry changed to a defined value,
+ * there's a need to update the CollectibleContract object in state.
+ *
+ * @param newCollectibleContract - CollectibleContract data object.
+ * @param oldCollectibleContract - CollectibleContract object to compare with.
+ * @returns Whether there are differences.
+ */
+export function compareCollectibleContract(
```
Comment (mcmire) [inline src/assets/assetsUtil.ts:51]: This seems to be an equality check. What do you think about inverting this and calling this `isCollectibleContractEqual`?

File: src/assets/assetsUtil.ts [hunk 3]
```diff
@@ -34,3 +38,34 @@ export function compareCollectiblesMetadata(
...
+  newCollectibleContract: CollectibleContract,
+  oldCollectibleContract: CollectibleContract,
+) {
+  const keys: (keyof CollectibleContract)[] = [
```
Comment (mcmire) [inline src/assets/assetsUtil.ts:55]: What do you think about moving this outside of the function?

File: src/assets/CollectiblesController.ts [hunk 4]
```diff
@@ -675,6 +671,26 @@ export class CollectiblesController extends BaseController<
...
+
+      if (existingEntry) {
+        if (compareCollectibleContract(newEntry, existingEntry)) {
+          const indexToRemove = collectibleContracts.findIndex(
+            (collectibleContract) =>
+              collectibleContract.address.toLowerCase() ===
+              address.toLowerCase(),
+          );
+          if (indexToRemove !== -1) {
+            collectibleContracts.splice(indexToRemove, 1);
```
Comment (mcmire) [inline src/assets/CollectiblesController.ts:687]: This potentially changes `allCollectibleContracts` in state. What do you think about making a copy of `collectibleContracts` and removing this fron the copy?

Comment (adonesky1) [reply src/assets/CollectiblesController.ts]: done here: https://github.com/MetaMask/controllers/pull/680/commits/2bef09890f6df2392729b8cf4b99623af9322c0f

File: src/assets/assetsUtil.ts (hunk 3 above)
Comment (adonesky1) [reply src/assets/assetsUtil.ts]: done here: https://github.com/MetaMask/controllers/pull/680/commits/2bef09890f6df2392729b8cf4b99623af9322c0f

File: src/assets/assetsUtil.ts (hunk 2 above)
Comment (adonesky1) [reply src/assets/assetsUtil.ts]: done here: https://github.com/MetaMask/controllers/pull/680/commits/2bef09890f6df2392729b8cf4b99623af9322c0f

File: src/assets/CollectiblesController.ts [hunk 5]
```diff
@@ -559,24 +562,19 @@ export class CollectiblesController extends BaseController<
...
-          );
-          /* istanbul ignore next */
-          if (indexToRemove !== -1) {
-            collectibles.splice(indexToRemove, 1);
-          }
-        } else {
+        if (isCollectibleMetadataEqual(collectibleMetadata, existingEntry)) {
           return collectibles;
         }
+        // TODO: Switch to indexToUpdate
```
Comment (gantunesr) [inline src/assets/CollectiblesController.ts:568]: This comment has been here for a while now. Are we going to do this change in the future or is it ok to remove it?

Comment (adonesky1) [reply src/assets/CollectiblesController.ts]: addressed here: https://github.com/MetaMask/controllers/pull/680/commits/21847f8049686b1189a5dc9da36b3d69a31d2bdf

**Comment by adonesky1** [comment]
@wachunei @mcmire & @gantunesr made a few more refactor changes in my latest commit: https://github.com/MetaMask/controllers/pull/680/commits/21847f8049686b1189a5dc9da36b3d69a31d2bdf

**Comment by adonesky1** [comment]
[Asked for some input from product](https://consensys.slack.com/archives/G01PKA4QC3C/p1643408333805459) on this topic

**Comment by Gudahtt** [comment]
It seems that we still need a product decision about how to handle updates and which data to keep. I'll move this into draft until we come to a conclusion on that

**Comment by adonesky1** [comment]
very stale

**Pull request closed without merging**